realm/realm-library/src/main/cpp/util.hpp

```diff
@@ -58,11 +61,73 @@ JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, void *reserved);
 #define STRINGIZE_DETAIL(x) #x
 #define STRINGIZE(x) STRINGIZE_DETAIL(x)
 
-// Exception handling
-#define CATCH_STD() \
-    catch (...) { \
-        ConvertException(env, __FILE__, __LINE__); \
+void ConvertException(JNIEnv* env);
+
+// Exception handling using lambdas
+// Credit: http://stackoverflow.com/questions/29202377/how-do-i-force-the-compiler-to-recognize-lambda-type-in-templated-function
+template<typename T, typename F>
+T try_catch(JNIEnv* env, F func) {
+    try {
+        // if an exception already has occurred, it is not possible to call any JNI
+        // functions, and we don't wish to try to execute the JNI function
+        // alternatively, we could clear the exception but then we can easily miss
+        // an error stage
+        if (env->ExceptionCheck() == JNI_FALSE) {
```

Comment:
I think this checking is not necessary. If this ever happens, then we have a bug in our code, and an exception will be thrown anyway. It doesn't worth to check it all the time especially in some very frequent calls.